### DISCUSSION TITLE
Backup private repository from Gitlab which is not owned by me

### DISCUSSION THREAD

**Discussion by lbdr** [cooperspencer/gickup#260] [Q&A] [answered]
I'd like to include a backup of a private repository from Gitlab where I am only a maintainer but not the owner of said repository using a token.

As far as I understand this following MWE should backup all repositories from Gitlab including the repositories not owned by me.

```yaml
source:
  gitlab:
    - token: foobar
```

I tested this for GitHub (with a other token of course) and all worked fine.
It seems like `gickup` can't list repositories not owned by the user directly.

Thank you all for your help. @cooperspencer keep up the good work!

**Comment by cooperspencer** [comment]
You are correct. I'll investigate on how to backup projects were you are not the owner but are a maintainer.
In the meantime you could try to create a backup like this:
```yaml
source:
  gitlab:
    - token: foobar
      user: owner-of-the-project
      include: ["your-project"]
    
```

**Reply by cooperspencer** [reply]
Actually I have to correct myself. I Just created a repository in Gitlab with a different user and gave my main user maintainer permissions. I just ran a backup for Gitlab and it cloned said repository too.

**Comment by lbdr** [comment]
Thanks for your quick response! Unfortunately I cannot confirm that.

For the overview and to recap, the starting point is the following.
![screenshot](https://github.com/user-attachments/assets/82f5dc80-eea0-4322-a6e0-fa1d4dd63bd0)

For the token I accepted all scopes.
```yaml
source:
  gitlab:
    - token: foobar
destination:
  local:
    - path: ~/testing
      structured: true
      zip: false
      bare: true
```

Log:
```
$ ./gickup 
2024-08-28T12:10:01+02:00 INF Reading conf.yml file=conf.yml
2024-08-28T12:10:01+02:00 INF Configuration loaded destinations=1 pairs=1 sources=1
2024-08-28T12:10:01+02:00 INF Backup run starting
2024-08-28T12:10:01+02:00 INF grabbing repositories from xyz stage=gitlab url=https://gitlab.com
2024-08-28T12:10:03+02:00 INF starting backup for https://gitlab.com/xyz/testing.git stage=backup
2024-08-28T12:10:03+02:00 INF cloning gitlab.com/xyz/testing.git path=/home/foo/testing stage=locally
2024-08-28T12:10:04+02:00 INF Backup run complete duration=3.154463547s
```

Result:
```
$ tree -L 3 ~/testing 
/home/foo/testing
└── gitlab.com
    └── xyz
        └── testing.git
```
The repository o... is missing.

Then I changed the conf as you suggested above.
```yaml
source:
  gitlab:
    - token: foobar
    - token: foobar
      user: owner-of-the-project
      include: ["o..."]
destination:
  local:
    - path: ~/testing
      structured: true
      zip: false
      bare: true
```

Again the log:
```
$ ./gickup --debug
2024-08-28T12:25:20+02:00 INF Reading conf.yml file=conf.yml
2024-08-28T12:25:20+02:00 INF Configuration loaded destinations=1 pairs=2 sources=2
2024-08-28T12:25:20+02:00 INF Backup run starting
2024-08-28T12:25:20+02:00 INF grabbing repositories from xyz stage=gitlab url=https://gitlab.com
2024-08-28T12:25:21+02:00 INF grabbing repositories from owner-of-the-project stage=gitlab url=https://gitlab.com
2024-08-28T12:25:22+02:00 INF starting backup for https://gitlab.com/xyz/testing.git stage=backup
2024-08-28T12:25:22+02:00 INF cloning gitlab.com/xyz/testing.git path=/home/foo/testing stage=locally
2024-08-28T12:25:24+02:00 INF Backup run complete duration=4.335697946s
```

As you can see there is no second cloning in the log. Tree looks as above.
Is there a way to get more debug information? Is there a way to forbid cloning over API on Gitlab or anything I've overseen?
Thanks for your help!

**Reply by cooperspencer** [reply]
I just tested it again and for me it works fine. I created a token with the `read_api` permissions.

I'll just paste everything I did down below:
![screenshot-2024-08-28-10-50-42](https://github.com/user-attachments/assets/f82de9f5-a9c1-4586-af4f-8ff7de697d35)
As you can see, I am just a mainainer in the second repo.

![screenshot-2024-08-28-10-51-22](https://github.com/user-attachments/assets/a5b07f7a-be2d-4a18-be8a-116bcff68b54)
The token only has the `read_api` permission and nothing else.

![screenshot-2024-08-28-10-53-05](https://github.com/user-attachments/assets/95163fa1-9b93-496e-acca-4b395e5563ef)
It just cloned the two repos.

It seems that I forgot to add a whole lot of debug outputs to gitlab. I'll add that and post the build pipeline for you, so we can get to the bottom of this issue.

**Reply by cooperspencer** [reply]
after the build is done, you can download the artifacts. 
https://github.com/cooperspencer/gickup/actions/runs/10595514411
This will actually output which repositories are found and if they are user or group repositories.

**Comment by lbdr** [comment]
In the meantime I created a second user on Gitlab for testing and I found out why.
Only repositories which belong to a group are cloned if the user is invited to the repository.

Repositories in Gitlab can be owned by a user under the namespace of the user.
In addition, repositories can be owned by a group under the namespace of the group.

The repository which I described above is owned by a user and not by a group and I'm a maintainer as well.
As said these repositories are not cloned by `gickup`.

I invited you on Gitlab to two testing repos so maybe you can confirm this behavoir.

**Reply by cooperspencer** [reply]
You are correct.
![screenshot-2024-08-28-13-25-32](https://github.com/user-attachments/assets/65f6c30a-24cd-4e77-8f63-59f7952deb9c)

**Reply by cooperspencer** [reply]
I found the issue. I know how to fix it, I just need a bit of time.

**Reply by lbdr** [reply]
No worries. Thanks!

**Reply by cooperspencer** [reply, marked as answer]
It should be fixed with this [build](https://github.com/cooperspencer/gickup/actions/runs/10601723333?pr=263).
Can you download the artifact, test it and get back to me?

**Reply by lbdr** [reply]
I tested the build and it works perfectly fine now!

**Reply by cooperspencer** [reply]
Perfect, thanks for testing. I just pushed a new release with the latest changes.
